propagation/w3c/src/test/java/brave/propagation/w3c/TraceContextPropagationTest.java

```diff
@@ -0,0 +1,123 @@
+import brave.propagation.TraceContextOrSamplingFlags;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.junit.Test;
+
+import static brave.internal.HexCodec.lowerHexToUnsignedLong;
+import static brave.propagation.Propagation.KeyFactory.STRING;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TraceContextPropagationTest {
+  Map<String, String> request = new LinkedHashMap<>();
+  Propagation<String> propagation = TraceContextPropagation.newFactory().create(STRING);
+  Injector<Map<String, String>> injector = propagation.injector(Map::put);
+  Extractor<Map<String, String>> extractor = propagation.extractor(Map::get);
+
+  TraceContext sampledContext = TraceContext.newBuilder()
+      .traceIdHigh(lowerHexToUnsignedLong("67891233abcdef01"))
+      .traceId(lowerHexToUnsignedLong("2345678912345678"))
+      .spanId(lowerHexToUnsignedLong("463ac35c9f6413ad"))
+      .sampled(true)
+      .build();
+  String validTraceparent = "00-67891233abcdef012345678912345678-463ac35c9f6413ad-01";
+  String validB3Single = "67891233abcdef012345678912345678-463ac35c9f6413ad-1";
+  String otherState = "congo=lZWRzIHRoNhcm5hbCBwbGVhc3VyZS4=";
+
+  @Test public void injects_b3_when_no_other_tracestate() {
+    sampledContext = sampledContext.toBuilder().addExtra(Tracestate.EMPTY).build();
+
+    injector.inject(sampledContext, request);
+
+    assertThat(request).containsEntry("tracestate", "b3=" + validB3Single);
+  }
+
+  @Test public void injects_b3_before_other_tracestate() {
+    Tracestate tracestate = Tracestate.create(otherState);
+
+    sampledContext = sampledContext.toBuilder().addExtra(tracestate).build();
+
+    injector.inject(sampledContext, request);
+
+    assertThat(request).containsEntry("tracestate", "b3=" + validB3Single + "," + otherState);
+  }
+
+  @Test public void extracts_b3_when_no_other_tracestate() {
+    request.put("traceparent", validTraceparent);
+    request.put("tracestate", "b3=" + validB3Single);
+
+    assertThat(extractor.extract(request)).isEqualTo(
+        TraceContextOrSamplingFlags.newBuilder(sampledContext).addExtra(Tracestate.EMPTY).build());
+  }
+
+  @Test public void extracts_b3_before_other_tracestate() {
+    request.put("traceparent", validTraceparent);
+    request.put("tracestate", "b3=" + validB3Single + "," + otherState);
+
+    Tracestate tracestate = Tracestate.create(otherState);
+
+    assertThat(extractor.extract(request)).isEqualTo(
+        TraceContextOrSamplingFlags.newBuilder(sampledContext).addExtra(tracestate).build());
+  }
+
+  @Test public void extracted_toString() {
+    request.put("traceparent", validTraceparent);
+    request.put("tracestate", "b3=" + validB3Single + "," + otherState);
+
+    assertThat(extractor.extract(request)).hasToString(
+        "Extracted{"
+            + "traceContext=" + sampledContext + ", "
+            + "samplingFlags=SAMPLED_REMOTE, "
+            + "extra=[tracestate: " + otherState + "]"
+            + "}");
+  }
```

Comment:
thanks to @anuraaga for all the advice, this is much more intuitive now, and especially useful for partially deconstructed formats like this.